tests/qam-kgraft/update_kgraft.pm

```diff
@@ -0,0 +1,164 @@
+# SUSE's openQA tests
+#
+# Copyright © 2016 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+use base 'opensusebasetest';
+use testapi;
+use qam;
+use utils;
+
+use strict;
+use warnings;
+
+sub timemark {
+    my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst) = localtime(time);
+    return sprintf("%02d%02d%02d.%02d%02d.%02d", $year % 100, $mon + 1, $mday, $hour, $min, $sec);
+}
+
+sub kgr_status {
+    # precompiled regexes
+    my $kgr_ready       = qr/^ready/;
+    my $kgr_in_progress = qr/^in_progress/;
+
+    script_run("kgr status | tee /dev/$serialdev", 0);
+    my $out = wait_serial([$kgr_ready, $kgr_in_progress], 10);
+    if ($out =~ $kgr_ready) {
+        return 1;
+    }
+    elsif ($out =~ $kgr_in_progress) {
+        return 0;
+    }
+}
+
+sub kgr_block {
+    my $kgr_block_free     = qr/kgr-$/;
+    my $kgr_block_hwrandom = qr/kgr-hwrng/;
+    my $kgr_block_other    = qr/kgr-.*/;
+    my $out                = "";
+
+    while (1) {
+        script_run("kgr poke");
+        script_run(qq{BLOCK="kgr-\$(kgr blocking)"; echo \$BLOCK |tee > /dev/$serialdev}, 0);
+        $out = wait_serial([$kgr_block_hwrandom, $kgr_block_other, $kgr_block_free], 30);
+
+        if ($out =~ $kgr_block_hwrandom) {
+            script_run("dd if=/dev/random of=/dev/null bs=2048 count=2000");
+        }
+        elsif ($out =~ $kgr_block_other) {
+        }
+        elsif ($out =~ $kgr_block_free) {
+            last;
+        }
+    }
+}
+
+sub run() {
+    my $self  = shift;
+    my $svirt = select_console('svirt');
+    my $name  = get_var('VIRSH_GUESTNAME');
+    $svirt->attach_to_running($name);
+    select_console('sut');
+    select_console("root-console");
+
+    capture_state("before");
+    script_run(qq{if \$(zypper lr | grep test-kgraft -q);then zypper rr test-kgraft ; fi });
+    #check kgr status
+    until (kgr_status) {
+        kgr_block;
+    }
+
+    # create reference snapshot
+    my $snapshot_before = "snap_before-" . timemark;
+    my $ret             = $svirt->run_cmd("virsh snapshot-create-as $name $snapshot_before");
+    die "snapshot $snapshot_before failed" if $ret;
+    set_var('KGRAFT_SNAPSHOT_BEFORE', $snapshot_before);
```

Comment:
as in before, this looks very backend specific and should not be part of the tests distribution